File: dsa/li19_bit/L389_FindTheDifference.java

public class L389_FindTheDifference {

	public static void main(String[] args) {
		L389_FindTheDifference obj = new L389_FindTheDifference();

		System.out.println(obj.findTheDifference("acd", "aedc"));
	}

	/*
	 * Problem: https://leetcode.com/problems/find-the-difference/
	 * Companies: Google
     * Idea: s and t has chars repeating. Hence do xor to eliminate all the repeating chars.
     * At the end, the odd char remains.
	 * */
	public char findTheDifference(String s, String t) {
        int chr = 0;
        for(char ch: s.toCharArray()){
            chr ^= ch;
        }
        for(char ch: t.toCharArray()){
            chr ^= ch;
        }

        return (char)chr;
    }

}

File: dsa/patterns/19_bit/python/l389_find_the_difference.py
# Problem: https://leetcode.com/problems/find-the-difference/
# Companies: Google
# Idea: s and t has chars repeating. Hence do xor to eliminate all the repeating chars.
# At the end, the odd char remains.
def find_the_difference(s, t):
    chr_val = 0
    for ch in s:
        chr_val ^= ord(ch)
    for ch in t:
        chr_val ^= ord(ch)

    return chr(chr_val)


# ------------------ Test (Same as Java main) ------------------

print(find_the_difference("acd", "aedc"))  # Expected: 'e'
